ext/strscan/strscan.c

```diff
@@ -709,19 +709,20 @@ strscan_do_scan(VALUE self, VALUE pattern, int succptr, int getstr, int headonly
     }
     else {
         StringValue(pattern);
-        rb_enc_check(p->str, pattern);
-        if (S_RESTLEN(p) < RSTRING_LEN(pattern)) {
-            return Qnil;
-        }
+        rb_encoding *enc = rb_enc_check(p->str, pattern);
 
         if (headonly) {
+            if (S_RESTLEN(p) < RSTRING_LEN(pattern)) {
+                return Qnil;
+            }
```

Comment:
Why do we need this move? Why is this needless for `!headonly`?

Reply:
Similar checks are made within `rb_memsearch()` within `!headonly`?

https://github.com/ruby/ruby/blob/cf8388f76c4c2ff2f46d0d2aa2cf5186e05ff606/re.c#L251-L256

```
long
rb_memsearch(const void *x0, long m, const void *y0, long n, rb_encoding *enc)
{
    const unsigned char *x = x0, *y = y0;

    if (m > n) return -1;
```

- m = `RSTRING_LEN(pattern)`
- n = `S_RESTLEN(p)`

This means the following : 
`if (RSTRING_LEN(pattern) > S_RESTLEN(p)) return -1;`